<a href="https://colab.research.google.com/github/yuraskass/lab1/blob/main/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [4]:
import pandas as pd

In [5]:
df = pd.read_csv("/content/d.csv")
df[:5]

loan_id   no_of_dependents      education  self_employed   income_annum  \
0        1                  2       Graduate             No        9600000   
1        2                  0   Not Graduate            Yes        4100000   
2        3                  3       Graduate             No        9100000   
3        4                  3       Graduate             No        8200000   
4        5                  5   Not Graduate            Yes        9800000   

    loan_amount   loan_term   cibil_score   residential_assets_value  \
0      29900000          12           778                    2400000   
1      12200000           8           417                    2700000   
2      29700000          20           506                    7100000   
3      30700000           8           467                   18200000   
4      24200000          20           382                   12400000   

    commercial_assets_value   luxury_assets_value   bank_asset_value  \
0                  17600000              22700000            8000000   
1                   2200000               8800000            3300000   
2                   4500000              33300000           12800000   
3                   3300000              23300000            7900000   
4                   8200000              29400000            5000000   

   loan_status  
0     Approved  
1     Rejected  
2     Rejected  
3     Rejected  
4     Rejected

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [6]:
df.columns

Index(['loan_id', ' no_of_dependents', ' education', ' self_employed',
       ' income_annum', ' loan_amount', ' loan_term', ' cibil_score',
       ' residential_assets_value', ' commercial_assets_value',
       ' luxury_assets_value', ' bank_asset_value', ' loan_status'],
      dtype='object')

In [7]:
df.head()

loan_id   no_of_dependents      education  self_employed   income_annum  \
0        1                  2       Graduate             No        9600000   
1        2                  0   Not Graduate            Yes        4100000   
2        3                  3       Graduate             No        9100000   
3        4                  3       Graduate             No        8200000   
4        5                  5   Not Graduate            Yes        9800000   

    loan_amount   loan_term   cibil_score   residential_assets_value  \
0      29900000          12           778                    2400000   
1      12200000           8           417                    2700000   
2      29700000          20           506                    7100000   
3      30700000           8           467                   18200000   
4      24200000          20           382                   12400000   

    commercial_assets_value   luxury_assets_value   bank_asset_value  \
0                  17600000              22700000            8000000   
1                   2200000               8800000            3300000   
2                   4500000              33300000           12800000   
3                   3300000              23300000            7900000   
4                   8200000              29400000            5000000   

   loan_status  
0     Approved  
1     Rejected  
2     Rejected  
3     Rejected  
4     Rejected

In [8]:
df[len(df)-5:len(df)]

df.tail()

loan_id   no_of_dependents      education  self_employed   income_annum  \
4264     4265                  5       Graduate            Yes        1000000   
4265     4266                  0   Not Graduate            Yes        3300000   
4266     4267                  2   Not Graduate             No        6500000   
4267     4268                  1   Not Graduate             No        4100000   
4268     4269                  1       Graduate             No        9200000   

       loan_amount   loan_term   cibil_score   residential_assets_value  \
4264       2300000          12           317                    2800000   
4265      11300000          20           559                    4200000   
4266      23900000          18           457                    1200000   
4267      12800000           8           780                    8200000   
4268      29700000          10           607                   17800000   

       commercial_assets_value   luxury_assets_value   bank_asset_value  \
4264                    500000               3300000             800000   
4265                   2900000              11000000            1900000   
4266                  12400000              18100000            7300000   
4267                    700000              14100000            5800000   
4268                  11800000              35700000           12000000   

      loan_status  
4264     Rejected  
4265     Approved  
4266     Rejected  
4267     Approved  
4268     Approved

In [9]:
df.describe()

loan_id   no_of_dependents   income_annum   loan_amount  \
count  4269.000000        4269.000000   4.269000e+03  4.269000e+03   
mean   2135.000000           2.498712   5.059124e+06  1.513345e+07   
std    1232.498479           1.695910   2.806840e+06  9.043363e+06   
min       1.000000           0.000000   2.000000e+05  3.000000e+05   
25%    1068.000000           1.000000   2.700000e+06  7.700000e+06   
50%    2135.000000           3.000000   5.100000e+06  1.450000e+07   
75%    3202.000000           4.000000   7.500000e+06  2.150000e+07   
max    4269.000000           5.000000   9.900000e+06  3.950000e+07   

         loan_term   cibil_score   residential_assets_value  \
count  4269.000000   4269.000000               4.269000e+03   
mean     10.900445    599.936051               7.472617e+06   
std       5.709187    172.430401               6.503637e+06   
min       2.000000    300.000000              -1.000000e+05   
25%       6.000000    453.000000               2.200000e+06   
50%      10.000000    600.000000               5.600000e+06   
75%      16.000000    748.000000               1.130000e+07   
max      20.000000    900.000000               2.910000e+07   

        commercial_assets_value   luxury_assets_value   bank_asset_value  
count              4.269000e+03          4.269000e+03       4.269000e+03  
mean               4.973155e+06          1.512631e+07       4.976692e+06  
std                4.388966e+06          9.103754e+06       3.250185e+06  
min                0.000000e+00          3.000000e+05       0.000000e+00  
25%                1.300000e+06          7.500000e+06       2.300000e+06  
50%                3.700000e+06          1.460000e+07       4.600000e+06  
75%                7.600000e+06          2.170000e+07       7.100000e+06  
max                1.940000e+07          3.920000e+07       1.470000e+07

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





1.   2135.000000	2.498712	5.059124e+06	1.513345e+07	10.900445	599.936051	7.472617e+06	4.973155e+06	1.512631e+07	4.976692e+06- средние


2.   1.000000	0.000000	2.000000e+05	3.000000e+05	2.000000	300.000000	-1.000000e+05	0.000000e+00	3.000000e+05	0.000000e+00-MIN


4269.000000	5.000000	9.900000e+06	3.950000e+07	20.000000	900.000000	2.910000e+07	1.940000e+07	3.920000e+07	1.470000e+07-MAX



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [59]:
"""

df.loan_id = df.loan_id.astype("category")
df.no_of_dependents = df.no_of_dependents.astype("category")
df.income_annum = df.income_annum.astype("int64")
df.loan_term = df.loan_term.astype("category")


"""
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4269 entries, 0 to 4268
Data columns (total 13 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   loan_id                    4269 non-null   int64 
 1    no_of_dependents          4269 non-null   int64 
 2    education                 4269 non-null   object
 3    self_employed             4269 non-null   object
 4    income_annum              4269 non-null   int64 
 5    loan_amount               4269 non-null   int64 
 6    loan_term                 4269 non-null   int64 
 7    cibil_score               4269 non-null   int64 
 8    residential_assets_value  4269 non-null   int64 
 9    commercial_assets_value   4269 non-null   int64 
 10   luxury_assets_value       4269 non-null   int64 
 11   bank_asset_value          4269 non-null   int64 
 12   loan_status               4269 non-null   object
dtypes: int64(10), object(3)
memory usage: 433.7+ KB


Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [64]:
df.loc[df['loan_id']<500]

loan_id   no_of_dependents      education  self_employed   income_annum  \
0          1                  2       Graduate             No        9600000   
1          2                  0   Not Graduate            Yes        4100000   
2          3                  3       Graduate             No        9100000   
3          4                  3       Graduate             No        8200000   
4          5                  5   Not Graduate            Yes        9800000   
..       ...                ...            ...            ...            ...   
494      495                  5   Not Graduate             No        2300000   
495      496                  2   Not Graduate            Yes        2400000   
496      497                  5       Graduate            Yes        9800000   
497      498                  2   Not Graduate             No        8100000   
498      499                  3       Graduate             No        6700000   

      loan_amount   loan_term   cibil_score   residential_assets_value  \
0        29900000          12           778                    2400000   
1        12200000           8           417                    2700000   
2        29700000          20           506                    7100000   
3        30700000           8           467                   18200000   
4        24200000          20           382                   12400000   
..            ...         ...           ...                        ...   
494       5700000          18           329                    5600000   
495       5400000          12           739                     300000   
496      30300000           6           867                   20800000   
497      24800000           6           307                     200000   
498      15500000          18           492                    2100000   

      commercial_assets_value   luxury_assets_value   bank_asset_value  \
0                    17600000              22700000            8000000   
1                     2200000               8800000            3300000   
2                     4500000              33300000           12800000   
3                     3300000              23300000            7900000   
4                     8200000              29400000            5000000   
..                        ...                   ...                ...   
494                   3300000               4900000            2700000   
495                   2500000               5000000            3000000   
496                  16000000              30200000            7600000   
497                   4000000              31100000           10500000   
498                   2000000              26300000            9400000   

     loan_status  
0       Approved  
1       Rejected  
2       Rejected  
3       Rejected  
4       Rejected  
..           ...  
494     Rejected  
495     Approved  
496     Approved  
497     Rejected  
498     Rejected  

[499 rows x 13 columns]

In [66]:
df.loc[df['loan_term']==12]

KeyError: ignored

In [67]:
df.loc[df['cibil_score']>500]


KeyError: ignored

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [53]:
group_df = df.groupby('loan_id').mean()
group_df

<ipython-input-53-657c55b0117f>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  group_df = df.groupby('loan_id').mean()


no_of_dependents   income_annum   loan_amount   loan_term  \
loan_id                                                               
1                      2.0      9600000.0    29900000.0        12.0   
2                      0.0      4100000.0    12200000.0         8.0   
3                      3.0      9100000.0    29700000.0        20.0   
4                      3.0      8200000.0    30700000.0         8.0   
5                      5.0      9800000.0    24200000.0        20.0   
...                    ...            ...           ...         ...   
4265                   5.0      1000000.0     2300000.0        12.0   
4266                   0.0      3300000.0    11300000.0        20.0   
4267                   2.0      6500000.0    23900000.0        18.0   
4268                   1.0      4100000.0    12800000.0         8.0   
4269                   1.0      9200000.0    29700000.0        10.0   

          cibil_score   residential_assets_value   commercial_assets_value  \
loan_id                                                                      
1               778.0                  2400000.0                17600000.0   
2               417.0                  2700000.0                 2200000.0   
3               506.0                  7100000.0                 4500000.0   
4               467.0                 18200000.0                 3300000.0   
5               382.0                 12400000.0                 8200000.0   
...               ...                        ...                       ...   
4265            317.0                  2800000.0                  500000.0   
4266            559.0                  4200000.0                 2900000.0   
4267            457.0                  1200000.0                12400000.0   
4268            780.0                  8200000.0                  700000.0   
4269            607.0                 17800000.0                11800000.0   

          luxury_assets_value   bank_asset_value  
loan_id                                           
1                  22700000.0          8000000.0  
2                   8800000.0          3300000.0  
3                  33300000.0         12800000.0  
4                  23300000.0          7900000.0  
5                  29400000.0          5000000.0  
...                       ...                ...  
4265                3300000.0           800000.0  
4266               11000000.0          1900000.0  
4267               18100000.0          7300000.0  
4268               14100000.0          5800000.0  
4269               35700000.0         12000000.0  

[4269 rows x 9 columns]

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.